### ISSUE TITLE
Maxwell does not setPosition in some situations, leading to errors

### ISSUE THREAD

**Issue by leonjoyce** [zendesk/maxwell#884]
In normal situation, maxwell captures an event from binlog, send it to producer then use context.setPosition to update binlog position in maxwell.position table.
However in the following cases maxwell does not update the position table:
1 For any DDL event.
2 if the filter include some dbs/tables, maxwell will only set position for the filtered tables/dbs.
The consequences of 1&2 are that maxwell's position is sometime behind the latest binlog, and in some cloud services like Aliyun, the binlog file maxwell positioning to are often purged, leading to maxwell reported errors(fail to read binlog file).

So I proposed some fixes:
For 1, when maxwell send DDL event to producer, it should also set positions.
For 2,  maxwell may keep setting positions only for filtered tables. However when maxwell restart, if it find the binlog is purged (it is positioning to a binlog file which is offline) , it should look for the latest binlog file.

I am new to maxwell community. Would like to hear comments on the fixes above. Thanks.

**Comment by leonjoyce** [comment]
Any comments?

**Comment by osheroff** [comment]
hmmmm ok.  I can see how this could be an issue in very low-traffic or heavily filtered scenarios.  

I think the proper thing to do to prevent purge errors is to maybe force a heartbeat out after we receive a ROTATE-BINLOG event.  This should cause the position to get bumped.  lms if I can make that work.

**Comment by leonjoyce** [comment]
@osheroff Do you have plans to implement the solution above?

**Comment by osheroff** [comment]
Yeah.  I need to grab some time but I will see abou he getting a fix together.

**Comment by osheroff** [comment]
@leonjoyce I went and looked at some code about this one, and I can't really get a reproduction together.  Any time maxwell sees that the binlog changes, it sends out a heartbeat (by writing to maxwell.heartbeats).  This heartbeat will always be recorded in `maxwell.positions`.  So even if there's filtered tables,  it should go like:

1.   Aliyun rotates log files
2.  Maxwell moves the binlog pointer ahead
3.  Maxwell sends a heartbeat into the current binlog file
4.  maxwell picks up that heartbeat and sets the position in the new file.

Now it seems that *something* in that chain isn't working properly, but I'm not sure what.  The fix I had thought might work won't.  Do you have further information on your issue?

**Comment by leonjoyce** [comment]
I met the same issue with #847.
Yes the chain you mention is not working.
my proposes fix is that , when maxwell detect missing binlogs, it should point to the latest binlog (using show master status), instead of repeatedly reboot. Agree with me?

**Comment by leonjoyce** [comment]
Maxwell was up when the logs rotated, leading to "Could not find first log file name..." error.
If the error happen, as you mention in issue847, we should drop database maxwell and re-point to the latest binlog position. 
But I think the alternative is to recapture schema and re-point to the latest binlog position, correct?
@osheroff

**Comment by osheroff** [comment]
@leonjoyce yes that's a possibility but really what I want to do is find and fix the issue that caused maxwell to somehow miss the log-rotation event.  Do you have logs from that event?  I doubt that they say much but I'm trying to figure out what might have happened to cause maxwell to miss the event.

To be clear:  even with table filtering/everything on, what should happen is:

1. logs rotate.  maxwell follows this rotate
2. maxwell notes that the file has changed and sends a heartbeat
3. the heartbeat event is picked up by the replicator and sent to the producer, which doesn't output anything but goes ahead and stores the new position anyway.

What producer are you using?  We've kind of left the implementation of (3) up to each producer and maybe yours isn't doing things right?  bah.

**Comment by leonjoyce** [comment]
I am using the Kafka producer

**Comment by leonjoyce** [comment]
com.zendesk.maxwell.replication.BinlogConnectorReplicator.getRow()case ROTATE:
					tableCache.clear();
					if ( stopOnEOF && event.getPosition().getOffset() > 0 ) {
						this.binlogEventListener.mustStop.set(true);
						this.client.disconnect();
						this.hitEOF = true;
						return null;
					}
					break;

**Comment by osheroff** [comment]
@leonjoyce this doesn't explicitly send the heartbeat, but you can see the logic here:

https://github.com/zendesk/maxwell/blob/387688019300d22948604bd07ada7c9a8bbace8b/src/main/java/com/zendesk/maxwell/schema/PositionStoreThread.java#L77

so even if maxwell doesn't see *any* events, we will still heartbeat every 10 seconds, and a heartbeat should cause the position to bump.  Leon, do you not see `maxwell.positions` increasing every 10 seconds?

**Comment by osheroff** [comment]
I'll be damned.  looks like table filters are preventing heartbeats from being read.

**Comment by nitingarg1** [comment]
Hi All,

The same issue occurs for me, Maxwell was up and running from last 4 weeks, but there was no activity in the Database for 2 weeks and then we start seeing the below error.

com.github.shyiko.mysql.binlog.network.ServerException: Could not find first log file name in binary log index file
	at com.github.shyiko.mysql.binlog.BinaryLogClient.listenForEventPackets(BinaryLogClient.java:882) [mysql-binlog-connector-java-0.16.1.jar:0.16.1]
	at com.github.shyiko.mysql.binlog.BinaryLogClient.connect(BinaryLogClient.java:559) [mysql-binlog-connector-java-0.16.1.jar:0.16.1]
	at com.github.shyiko.mysql.binlog.BinaryLogClient$7.run(BinaryLogClient.java:793) [mysql-binlog-connector-java-0.16.1.jar:0.16.1]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_191]
18:17:14,174 INFO  BinlogConnectorLifecycleListener - Binlog disconnected.
18:17:14,235 WARN  BinlogConnectorReplicator - replicator stopped at position: mysql-bin-changelog.009119:67353560 -- restarting
18:17:14,399 INFO  BinaryLogClient - Connected to zuni-test-db-cluster.cluster-cav1zghpl4m8.us-east-1.rds.amazonaws.com:3306 at mysql-bin-changelog.009119/67353560 (sid:6379, cid:623538)
18:17:14,399 INFO  BinlogConnectorLifecycleListener - Binlog connected.
18:17:14,438 WARN  BinlogConnectorLifecycleListener - Communication failure.
com.github.shyiko.mysql.binlog.network.ServerException: Could not find first log file name in binary log index file
	at com.github.shyiko.mysql.binlog.BinaryLogClient.listenForEventPackets(BinaryLogClient.java:882) [mysql-binlog-connector-java-0.16.1.jar:0.16.1]
	at com.github.shyiko.mysql.binlog.BinaryLogClient.connect(BinaryLogClient.java:559) [mysql-binlog-connector-java-0.16.1.jar:0.16.1]
	at com.github.shyiko.mysql.binlog.BinaryLogClient$7.run(BinaryLogClient.java:793) [mysql-binlog-connector-java-0.16.1.jar:0.16.1]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_191]

**Comment by leonjoyce** [comment]
The solution above only works for scenario when maxwell schema DB is the same as replication DB. 
If they are different, then the position will not bump.
@osheroff please confirm.

**Comment by TengHuo** [comment]
Hi @leonjoyce @osheroff 
Thank you so much for the explanation about this issue. It helps me a lot.

We got the same issue like @nitingarg1 , and we are running Maxwell with a separate metadata db, it read replication binlog from a MySQL server, and save Maxwell db in another MySQL server. So this pr #960 didn't cover this kind of situation

* Binary logs rotate. Maxwell follows this rotate, but we setup a lot of filters, so only few data was send to Producer (we are using Kafka),
* Positions update infrequently, and because logs rotate, so there is no heartbeat,
* binlog was purged due to MySQL purged rule from DBA,
* Maxwell try to restart from an old position which was already purged, so it can't run anymore before we clean Maxwell metadata db.

I read the code of Maxwell, and I noticed there is a function named `suppress()` in `RowMap.java`. The row is marked as "suppressed" won't be sent to a Producer, but it will update positions.

My question is if we can fix this issue (position not update) by marking a row as  "suppressed" when we filter it in `BinlogConnectorReplicator` rather than discard it?
